xarray/core/types.py

```diff
@@ -156,8 +156,15 @@ def copy(
 
 T_DataArrayOrSet = TypeVar("T_DataArrayOrSet", bound=Union["Dataset", "DataArray"])
 
-# Maybe we rename this to T_Data or something less Fortran-y?
+# Use `T_Xarray` for most functions, when the return type matches the input type — for
+# example `ds.where(cond)` returns a `Dataset` when `ds` is a `Dataset`, and
+# `da.where(cond)` returns a `DataArray` when `da` is a `DataArray`.
 T_Xarray = TypeVar("T_Xarray", "DataArray", "Dataset")
+
+# Use `T_DataWithCoords` for:
+# - functions that return either `DataArray` or `Dataset`, and
```

Comment:
I think for this usecase you should use `T_DataArrayOrSet`.
You can use this one, when the code that uses the input only uses methods that are available in `DataWithCoords`.

Reply:
Yes OK. I had tried to consolidate in #8213. 

Is the rule for what to use really "the code that uses the input only uses methods that are available in DataWithCoords" though? And then if we add an extra method, we change everything to use `T_DataArrayOrSet`, given they're not always compatible?

(Maybe this isn't an easy answer, and I appreciate your review)

Reply:
I tried another pass. If it's not quite there yet, feel free to just rewrite and/or push off for the moment